File: libs/python/mail_runtime_core/runtime.py
#!/usr/bin/env python3
"""Shared mail pipeline runtime: envelopes, rules, actions, and dispatch."""

from __future__ import annotations

import re
import shutil
import tempfile
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, Protocol


@dataclass
class AttachmentMeta:
    """Metadata about a message attachment."""

    name: str
    content_type: str
    is_inline: bool = False
    content_id: str | None = None


@dataclass
class MailEnvelope:
    """Normalized message shape used by sources, rules, and actions."""

    message_id: str
    provider: str
    account_id: str
    mailbox_id: str | None
    sender_name: str
    sender_email: str
    subject: str
    received_at: str | None = None
    body_text: str | None = None
    body_html: str | None = None
    headers: dict[str, str] = field(default_factory=dict)
    has_attachments: bool = False
    raw: dict[str, Any] = field(default_factory=dict)


@dataclass
class ActionResult:
    """Structured side effect emitted by an action."""

    kind: str
    payload: dict[str, Any]


class MailProviderClient(Protocol):
    """Minimal provider surface actions can rely on."""

    def fetch_body(self, envelope: MailEnvelope) -> MailEnvelope: ...

    def list_attachments(self, envelope: MailEnvelope) -> list[AttachmentMeta]: ...

    def download_attachments(
        self,
        envelope: MailEnvelope,
        output_dir: str,
        *,
        content_types: list[str] | None = None,
        inline_only: bool | None = None,
        include_body_html: bool = False,
    ) -> list[str]: ...


@dataclass
class ActionContext:
    """Runtime context passed to action handlers."""

    envelope: MailEnvelope
    provider_client: MailProviderClient
    workspace: Path
    logger: Callable[[str], None]
    config: dict[str, Any] = field(default_factory=dict)
    artifacts: dict[str, Any] = field(default_factory=dict)


@dataclass
class RegisteredAction:
    """Action handler plus declared data requirements."""

    name: str
    handler: Callable[[ActionContext, dict[str, Any]], list[ActionResult]]
    needs_body: bool = False
    attachment_request: dict[str, Any] | None = None


def normalize_action(action: str | dict[str, Any]) -> tuple[str, dict[str, Any]]:
    if isinstance(action, str):
        return action, {}
    return action["name"], action.get("params", {})


class ActionRegistry:
    """Registry and executor for named mail actions."""

    def __init__(self):
        self._actions: dict[str, RegisteredAction] = {}

    def register(
        self,
        name: str,
        handler: Callable[[ActionContext, dict[str, Any]], list[ActionResult]],
        *,
        needs_body: bool = False,
        attachment_request: dict[str, Any] | None = None,
    ) -> None:
        self._actions[name] = RegisteredAction(
            name=name,
            handler=handler,
            needs_body=needs_body,
            attachment_request=attachment_request,
        )

    def get(self, name: str) -> RegisteredAction:
        if name not in self._actions:
            raise KeyError(f"Unknown mail action: {name}")
        return self._actions[name]


def _to_list(value: Any) -> list[Any]:
    if isinstance(value, list):
        return value
    return [value]


def _matches_any_exact(actual: str, expected: Any) -> bool:
    actual_low = (actual or "").lower()
    return any(actual_low == str(item).lower() for item in _to_list(expected))


def _matches_any_contains(actual: str, expected: Any) -> bool:
    actual_low = (actual or "").lower()
    return any(str(item).lower() in actual_low for item in _to_list(expected))


def _matches_any_prefix(actual: str, expected: Any) -> bool:
    actual_low = (actual or "").lower()
    return any(actual_low.startswith(str(item).lower()) for item in _to_list(expected))


def _sender_domain(sender_email: str) -> str:
    if "@" not in (sender_email or ""):
        return ""
    return sender_email.split("@", 1)[1].lower()


def _matches_any_domain(actual: str, expected: Any) -> bool:
    actual_low = (actual or "").lower()
    for item in _to_list(expected):
        wanted = str(item).lower()
        if actual_low == wanted or actual_low.endswith("." + wanted):
            return True
    return False


def _body_text(envelope: MailEnvelope) -> str:
    return " ".join(filter(None, [envelope.body_text, envelope.body_html]))


def rule_matches(envelope: MailEnvelope, rule: dict[str, Any]) -> bool:
    """Return True when an envelope matches a deterministic rule."""

    if rule.get("providers") and not _matches_any_exact(envelope.provider, rule["providers"]):
        return False
    if rule.get("accounts") and not _matches_any_exact(envelope.account_id, rule["accounts"]):
        return False
    if rule.get("mailboxes") and not _matches_any_exact(envelope.mailbox_id or "", rule["mailboxes"]):
        return False

    match = rule.get("match", {})
    if not match:
        return True

    body = None
    for key, expected in match.items():
        if key == "sender_email":
            if not _matches_any_exact(envelope.sender_email, expected):
                return False
        elif key == "sender_domain":
            if not _matches_any_domain(_sender_domain(envelope.sender_email), expected):
                return False
        elif key == "sender_name_contains":
            if not _matches_any_contains(envelope.sender_name, expected):
                return False
        elif key == "subject":
            if not _matches_any_exact(envelope.subject, expected):
                return False
        elif key == "subject_contains":
            if not _matches_any_contains(envelope.subject, expected):
                return False
        elif key == "subject_prefix":
            if not _matches_any_prefix(envelope.subject, expected):
                return False
        elif key == "subject_regex":
            patterns = _to_list(expected)
            if not any(re.search(pattern, envelope.subject or "", re.IGNORECASE) for pattern in patterns):
                return False
        elif key == "body_contains":
            body = body if body is not None else _body_text(envelope)
            if not _matches_any_contains(body, expected):
                return False
        elif key == "has_attachments":
            if bool(expected) != bool(envelope.has_attachments):
                return False
        else:
            raise ValueError(f"Unsupported mail rule condition: {key}")

    return True


def select_matching_rules(envelope: MailEnvelope, rules: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Return matching rules in execution order."""

    matches = []
    for rule in rules:
        if not rule.get("enabled", True):
            continue
        if rule_matches(envelope, rule):
            matches.append(rule)
            if not rule.get("continue", False):
                break
    return matches


def execute_rules(
    envelope: MailEnvelope,
    rules: list[dict[str, Any]],
    registry: ActionRegistry,
    provider_client: MailProviderClient,
    *,
    workspace: Path,
    logger: Callable[[str], None],
    config: dict[str, Any] | None = None,
) -> tuple[list[dict[str, Any]], list[ActionResult]]:
    """Execute matching rule actions and collect structured results."""

    workspace.mkdir(parents=True, exist_ok=True)
    matched = select_matching_rules(envelope, rules)
    results: list[ActionResult] = []

    if matched:
        logger(
            "matched mail rule(s): "
            + ", ".join(rule.get("id", "<unnamed>") for rule in matched)
            + f" | sender={envelope.sender_email} | subject={envelope.subject}"
        )

    for rule in matched:
        for action_cfg in rule.get("actions", []):
            action_name, params = normalize_action(action_cfg)
            action = registry.get(action_name)
            logger(
                f"running mail action {action_name} for rule {rule.get('id', '<unnamed>')}"
            )
            ctx = ActionContext(
                envelope=envelope,
                provider_client=provider_client,
                workspace=workspace,
                logger=logger,
                config=config or {},
            )

            temp_dir: str | None = None
            if action.needs_body:
                ctx.envelope = provider_client.fetch_body(ctx.envelope)

            if action.attachment_request:
                temp_dir = tempfile.mkdtemp(prefix=f"mail-{action_name}-", dir=str(workspace))
                request = dict(action.attachment_request)
                downloaded = provider_client.download_attachments(
                    ctx.envelope,
                    temp_dir,
                    content_types=request.get("content_types"),
                    inline_only=request.get("inline_only"),
                    include_body_html=request.get("include_body_html", False),
                )
                ctx.artifacts["download_dir"] = temp_dir
                ctx.artifacts["downloaded_files"] = downloaded
                logger(
                    f"downloaded {len(downloaded)} artifact(s) for action {action_name}"
                )

            try:
                action_results = action.handler(ctx, params) or []
                results.extend(action_results)
            finally:
                if temp_dir and not params.get("keep_downloads"):
                    shutil.rmtree(temp_dir, ignore_errors=True)

    return matched, results

File: libs/ts/mail_runtime_core/src/runtime.ts
/**
 * Shared mail pipeline runtime: envelopes, rules, actions, and dispatch.
 * TS port of mail_runtime_core/runtime.py
 */

import { mkdirSync, mkdtempSync, rmSync } from "node:fs";
import { join } from "node:path";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AttachmentMeta {
  name: string;
  content_type: string;
  is_inline?: boolean;
  content_id?: string | null;
}

export interface AuthResults {
  /** "pass" | "fail" | "none" — result of DKIM verification */
  dkim?: string;
  /** "pass" | "fail" | "none" — result of SPF verification */
  spf?: string;
  /** "pass" | "fail" | "none" — result of DMARC verification */
  dmarc?: string;
  /** Raw Authentication-Results header value */
  raw?: string;
}

export interface MailEnvelope {
  message_id: string;
  provider: string;
  account_id: string;
  mailbox_id: string | null;
  sender_name: string;
  sender_email: string;
  subject: string;
  received_at?: string | null;
  body_text?: string | null;
  body_html?: string | null;
  headers?: Record<string, string>;
  has_attachments?: boolean;
  auth_results?: AuthResults;
  raw?: Record<string, unknown>;
}

export interface ActionResult {
  kind: string;
  payload: Record<string, unknown>;
}

export interface MailProviderClient {
  fetchBody(envelope: MailEnvelope): MailEnvelope | Promise<MailEnvelope>;
  listAttachments(envelope: MailEnvelope): AttachmentMeta[] | Promise<AttachmentMeta[]>;
  downloadAttachments(
    envelope: MailEnvelope,
    outputDir: string,
    options?: {
      content_types?: string[] | null;
      inline_only?: boolean | null;
      include_body_html?: boolean;
    },
  ): string[] | Promise<string[]>;
}

export interface ActionContext {
  envelope: MailEnvelope;
  provider_client: MailProviderClient;
  workspace: string;
  logger: (msg: string) => void;
  config: Record<string, unknown>;
  artifacts: Record<string, unknown>;
}

export interface RegisteredAction {
  name: string;
  handler: (ctx: ActionContext, params: Record<string, unknown>) => ActionResult[] | Promise<ActionResult[]>;
  needs_body: boolean;
  attachment_request: Record<string, unknown> | null;
}

// ---------------------------------------------------------------------------
// normalizeAction
// ---------------------------------------------------------------------------

export function normalizeAction(
  action: string | Record<string, unknown>,
): [string, Record<string, unknown>] {
  if (typeof action === "string") {
    return [action, {}];
  }
  return [action["name"] as string, (action["params"] as Record<string, unknown>) ?? {}];
}

// ---------------------------------------------------------------------------
// ActionPlugin
// ---------------------------------------------------------------------------

/**
 * Interface for external action plugin modules loaded dynamically at startup.
 * Any ESM module exporting a `register` function satisfies this interface.
 *
 * @example
 * // my-action-plugin/src/index.ts
 * import type { ActionPlugin, ActionRegistry } from '@openclaw/mail-runtime-core';
 * export const register: ActionPlugin['register'] = (registry) => {
 *   registry.register('my_custom_action', async (ctx, params) => { ... });
 * };
 */
export interface ActionPlugin {
  register(registry: ActionRegistry): void | Promise<void>;
}

// ---------------------------------------------------------------------------
// ActionRegistry
// ---------------------------------------------------------------------------

export class ActionRegistry {
  private _actions: Map<string, RegisteredAction> = new Map();

  register(
    name: string,
    handler: RegisteredAction["handler"],
    options?: { needs_body?: boolean; attachment_request?: Record<string, unknown> | null },
  ): void {
    this._actions.set(name, {
      name,
      handler,
      needs_body: options?.needs_body ?? false,
      attachment_request: options?.attachment_request ?? null,
    });
  }

  get(name: string): RegisteredAction {
    const action = this._actions.get(name);
    if (!action) {
      throw new Error(`Unknown mail action: ${name}`);
    }
    return action;
  }
}

// ---------------------------------------------------------------------------
// Rule matching helpers
// ---------------------------------------------------------------------------

function toList(value: unknown): unknown[] {
  if (Array.isArray(value)) return value;
  return [value];
}

function matchesAnyExact(actual: string, expected: unknown): boolean {
  const actualLow = (actual ?? "").toLowerCase();
  return toList(expected).some((item) => actualLow === String(item).toLowerCase());
}

function matchesAnyContains(actual: string, expected: unknown): boolean {
  const actualLow = (actual ?? "").toLowerCase();
  return toList(expected).some((item) => actualLow.includes(String(item).toLowerCase()));
}

function matchesAnyPrefix(actual: string, expected: unknown): boolean {
  const actualLow = (actual ?? "").toLowerCase();
  return toList(expected).some((item) => actualLow.startsWith(String(item).toLowerCase()));
}

function senderDomain(senderEmail: string): string {
  if (!(senderEmail ?? "").includes("@")) return "";
  return senderEmail.split("@", 2)[1].toLowerCase();
}

function matchesAnyDomain(actual: string, expected: unknown): boolean {
  const actualLow = (actual ?? "").toLowerCase();
  for (const item of toList(expected)) {
    const wanted = String(item).toLowerCase();
    if (actualLow === wanted || actualLow.endsWith("." + wanted)) return true;
  }
  return false;
}

function bodyText(envelope: MailEnvelope): string {
  return [envelope.body_text, envelope.body_html].filter(Boolean).join(" ");
}

// ---------------------------------------------------------------------------
// ruleMatches
// ---------------------------------------------------------------------------

export function ruleMatches(
  envelope: MailEnvelope,
  rule: Record<string, unknown>,
): boolean {
  if (rule["providers"] && !matchesAnyExact(envelope.provider, rule["providers"])) return false;
  if (rule["accounts"] && !matchesAnyExact(envelope.account_id, rule["accounts"])) return false;
  if (rule["mailboxes"] && !matchesAnyExact(envelope.mailbox_id ?? "", rule["mailboxes"])) return false;

  const match = (rule["match"] as Record<string, unknown>) ?? {};
  if (Object.keys(match).length === 0) return true;

  let body: string | null = null;

  for (const [key, expected] of Object.entries(match)) {
    switch (key) {
      case "sender_email":
        if (!matchesAnyExact(envelope.sender_email, expected)) return false;
        break;
      case "sender_domain":
        if (!matchesAnyDomain(senderDomain(envelope.sender_email), expected)) return false;
        break;
      case "sender_name_contains":
        if (!matchesAnyContains(envelope.sender_name, expected)) return false;
        break;
      case "subject":
        if (!matchesAnyExact(envelope.subject, expected)) return false;
        break;
      case "subject_contains":
        if (!matchesAnyContains(envelope.subject, expected)) return false;
        break;
      case "subject_prefix":
        if (!matchesAnyPrefix(envelope.subject, expected)) return false;
        break;
      case "subject_regex": {
        const patterns = toList(expected);
        if (!patterns.some((p) => new RegExp(String(p), "i").test(envelope.subject ?? "")))
          return false;
        break;
      }
      case "body_contains":
        if (body === null) body = bodyText(envelope);
        if (!matchesAnyContains(body, expected)) return false;
        break;
      case "has_attachments":
        if (Boolean(expected) !== Boolean(envelope.has_attachments)) return false;
        break;
      case "dkim_pass": {
        const dkimResult = (envelope.auth_results?.dkim ?? "none").toLowerCase();
        const wantPass = Boolean(expected);
        if (wantPass !== (dkimResult === "pass")) return false;
        break;
      }
      case "spf_pass": {
        const spfResult = (envelope.auth_results?.spf ?? "none").toLowerCase();
        const wantPass = Boolean(expected);
        if (wantPass !== (spfResult === "pass")) return false;
        break;
      }
      case "dmarc_pass": {
        const dmarcResult = (envelope.auth_results?.dmarc ?? "none").toLowerCase();
        const wantPass = Boolean(expected);
        if (wantPass !== (dmarcResult === "pass")) return false;
        break;
      }
      default:
        throw new Error(`Unsupported mail rule condition: ${key}`);
        break;
    }
  }

  return true;
}

// ---------------------------------------------------------------------------
// selectMatchingRules
// ---------------------------------------------------------------------------

export function selectMatchingRules(
  envelope: MailEnvelope,
  rules: Record<string, unknown>[],
): Record<string, unknown>[] {
  const matches: Record<string, unknown>[] = [];
  for (const rule of rules) {
    if ((rule["enabled"] ?? true) === false) continue;
    if (ruleMatches(envelope, rule)) {
      matches.push(rule);
      if (!rule["continue"]) break;
    }
  }
  return matches;
}

// ---------------------------------------------------------------------------
// executeRules
// ---------------------------------------------------------------------------

export async function executeRules(
  envelope: MailEnvelope,
  rules: Record<string, unknown>[],
  registry: ActionRegistry,
  providerClient: MailProviderClient,
  options: {
    workspace: string;
    logger: (msg: string) => void;
    config?: Record<string, unknown>;
  },
): Promise<[Record<string, unknown>[], ActionResult[]]> {
  const { workspace, logger, config } = options;
  mkdirSync(workspace, { recursive: true });
  const matched = selectMatchingRules(envelope, rules);
  const results: ActionResult[] = [];

  if (matched.length > 0) {
    logger(
      "matched mail rule(s): " +
        matched.map((r) => (r["id"] as string) ?? "<unnamed>").join(", ") +
        ` | sender=${envelope.sender_email} | subject=${envelope.subject}`,
    );
  }

  for (const rule of matched) {
    const actions = (rule["actions"] as unknown[]) ?? [];
    for (const actionCfg of actions) {
      const [actionName, params] = normalizeAction(actionCfg as string | Record<string, unknown>);
      const action = registry.get(actionName);
      logger(`running mail action ${actionName} for rule ${(rule["id"] as string) ?? "<unnamed>"}`);

      const ctx: ActionContext = {
        envelope: { ...envelope },
        provider_client: providerClient,
        workspace,
        logger,
        config: config ?? {},
        artifacts: {},
      };

      let tempDir: string | null = null;

      if (action.needs_body) {
        ctx.envelope = await providerClient.fetchBody(ctx.envelope);
      }

      if (action.attachment_request) {
        tempDir = mkdtempSync(join(workspace, `mail-${actionName}-`));
        const request = { ...action.attachment_request };
        const downloaded = await providerClient.downloadAttachments(ctx.envelope, tempDir, {
          content_types: (request["content_types"] as string[]) ?? undefined,
          inline_only: (request["inline_only"] as boolean) ?? undefined,
          include_body_html: (request["include_body_html"] as boolean) ?? false,
        });
        ctx.artifacts["download_dir"] = tempDir;
        ctx.artifacts["downloaded_files"] = downloaded;
        logger(`downloaded ${downloaded.length} artifact(s) for action ${actionName}`);
      }

      try {
        const actionResults = (await action.handler(ctx, params)) ?? [];
        results.push(...actionResults);
      } finally {
        if (tempDir && !params["keep_downloads"]) {
          rmSync(tempDir, { recursive: true, force: true });
        }
      }
    }
  }

  return [matched, results];
}
